src/openforms/forms/fd_translations_converter.py

```diff
@@ -0,0 +1,176 @@
+"""
+Implements the translation converter for form definitions.
+
+Github issue: https://github.com/open-formulieren/open-forms/issues/2958
+
+Before, translations for formio components were saved in the JSONField
+``FormDefinition.component_translations``, which is a key-value mapping of language
+code to translations mapping, which itself is a key-value mapping of source string to
+translation string. This had the drawback that if different components happen to use the
+same literal, they also shared the same translation.
+
+The utilities here help in converting these translations to component-local
+translations, where the translations are stored on each component inside the formio
+schema. The exact set of supported fields for translation varies per component, the
+reference for this is the typescript type definitions:
+https://open-formulieren.github.io/types/modules/index.html
+"""
+import logging
+from typing import TypedDict
+
+from glom import assign
+
+from openforms.formio.service import iter_components
+from openforms.formio.typing import Component
+
+logger = logging.getLogger(__name__)
+
+
+class Option(TypedDict):
+    value: str
+    label: str
+
+
+def _set_translations(
+    obj: Component | Option,
+    translations_from_store: dict[str, str],
+    locale: str,
+    fields: list[str],
+) -> None:
+    translations = {
+        key: translation
+        for key in fields
+        if (literal := obj.get(key))
+        and (translation := translations_from_store.get(literal))
+    }
+    if not translations:
+        return
+    assign(obj, f"openForms.translations.{locale}", translations, missing=dict)
+
+
+def _move_translations(component: Component, locale: str, translations: dict[str, str]):
+    """
+    Given a component and translation store, mutate the component to bake in translations.
+
+    This mutates the ``component`` parameter!
+    """
+    assert "type" in component
+
+    match component:
```

Comment:
Any way we could have the translatable fields set somewhere so that they can be retrieved for later use? Maybe not that relevant if this is the only place we will make use of them

Reply:
This is the only place - the frontend types basically define what is and isn't translatable and that should be sufficient.

The data structure of translations is crucial, since it is processed in a generic way in the backend, see: https://github.com/open-formulieren/open-forms/blob/master/src/openforms/formio/registry.py#L170

For components that have additional translations support, there is a hook, e.g.: https://github.com/open-formulieren/open-forms/blob/master/src/openforms/formio/components/vanilla.py#L178